crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -358,11 +358,60 @@ impl PgManager {
             .or(before)
             .map(|cursor| self.parse_tx_cursor(&cursor))
             .transpose()?;
-        if let Some(cursor) = cursor {
+
+        let mut after_tx_seq_num: Option<i64> = None;
+        let mut before_tx_seq_num: Option<i64> = None;
+        if let Some(filter) = &filter {
+            if let Some(checkpoint) = filter.after_checkpoint {
+                let subquery = transactions::dsl::transactions
+                    .filter(transactions::dsl::checkpoint_sequence_number.eq(checkpoint as i64))
+                    .order(transactions::dsl::tx_sequence_number.asc())
+                    .select(transactions::dsl::tx_sequence_number)
+                    .into_boxed();
+
+                after_tx_seq_num = self
+                    .run_query_async(|conn| subquery.get_result::<i64>(conn).optional())
+                    .await?;
+
+                // Return early if we cannot find txs after the specified checkpoint
+                if after_tx_seq_num.is_none() {
+                    return Ok(None);
+                }
+            }
+
+            if let Some(checkpoint) = filter.before_checkpoint {
+                let subquery = transactions::dsl::transactions
+                    .filter(transactions::dsl::checkpoint_sequence_number.eq(checkpoint as i64))
+                    .order(transactions::dsl::tx_sequence_number.desc())
+                    .select(transactions::dsl::tx_sequence_number)
+                    .into_boxed();
+
+                before_tx_seq_num = self
+                    .run_query_async(|conn| subquery.get_result::<i64>(conn).optional())
+                    .await?;
+
+                // Return early if we cannot find tx before the specified checkpoint
+                if before_tx_seq_num.is_none() {
+                    return Ok(None);
+                }
```

Comment:
Noting a future optimisation:  You can get the information you need in a single DB roundtrip as follows:

```sql
SELECT 
  MIN(tx_sequence_number) AS after,
  MAX(tx_sequence_number) + 1 AS before,
FROM
  transactions
WHERE
  checkpoint_sequence_number = "..."
GROUP BY
  checkoint_sequence_number;
```

But double check that this makes good use of the indices.

Reply:
We could probably also get this down to a single DB roundtrip as well for the whole query.